mautic/mautic
# Dynamic Content Channel

Please answer the following questions:

| Q             | A
| ------------- | ---
| Bug fix?      | N
| New feature?  | Y
| BC breaks?    | N
| Deprecations? | N
| Fixed issues  |  N/A

## Description

This PR provides the new Dynamic Content Channel. It allows you to define Dynamic Content in your campaign that will return different data based on your campaign flow. Needless to say, I'm pretty excited for this feature.

## Steps to test this PR

Apply the PR, then clear your cache. Refresh your admin and you'll see a new item under your Components menu called "Dynamic Content". Click that and you'll see the standard list view for channels. In the top right you are able to create new Dynamic Content items. Go ahead and create one and you'll then see it in the list. Create a few of them with alternate contents.

Now create a campaign using Campaign Forms as the source and open the Builder. Under Decisions on the right, you'll see a new "Request Dynamic Content" (RDC) button. Click or drag that to the canvas to add open the modal. Here you can enter a descriptive name, but more importantly a Slot Name must be entered. This is the name of the Dynamic Content Area (slot) that you want to populate on your website using one of our CMS plugins (not yet ready, working on tomorrow). So, for the sake of argument, enter `call_to_action_1` as your Slot Name. Then select the default Dynamic Content item that you want to show in case none of the "Push Dynamic Content" (PDC) actions (explained soon) are able to execute.

Once that's on your canvas, you can add in any number of conditions or actions just as normal. Go ahead and add a Form Field Value condition (whatever values you'd like) and attach the "Yes" side of the Request decision to the top of that. Now add a PDC action to the canvas, and select the variant content from the select list in the modal, and attach that to the "Yes" side of the Contact Field Value condition. Also create a PDC action and attach it to the "No" side of the contact field value. You now have a basic Dynamic Content setup as part of your campaign.

Now, open a new browser (or log out of Mautic), and create a new Contact in Mautic by visiting your website and filling out the form that's the entry point to the Campaign you created earlier (this way you're added to the Campaign). Once that's complete, open a new tab and navigate to `http://your_mautic_site.com/dwc/call_to_action_1`. This will trigger the campaign decision and return the appropriate content based on your campaign flow. Congratulations, you just tested it!
